### PR TITLE
First implementation for two endpoints.

### PR DESCRIPTION
**Pull request by sgmv**
First implementation for SD3.

### CODE REVIEW

**Review by artem-lifshits — APPROVED** [review]

**Review by vladimirvshivkov — APPROVED** [review]

File: internal/app/app.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/gin-gonic/gin"
+	"github.com/stackmon/otc-status-dashboard/internal/conf"
+	"github.com/stackmon/otc-status-dashboard/internal/db"
+	"go.uber.org/zap"
+	"net/http"
+)
+
+type App struct {
+	// Configuration
+	c *conf.Config
```
Comment (anton-sidelnikov) [inline internal/app/app.go:15]: don’t be stingy with variable namesб use ones that are immediately understandable: config for *conf.Config, router for *gin.Engine and so on

File: internal/app/middleware.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package app
+
+import (
+	"github.com/gin-gonic/gin"
+)
+
+func ErrorHandle() gin.HandlerFunc {
```
Comment (anton-sidelnikov) [inline internal/app/middleware.go:7]: i think better to make like this:
```
var (
	// ErrInternal is an error for when an internal service fails to process the request
	ErrInternal = errors.New("internal error")
	// ErrDataNotFound is an error for when requested data is not found
	ErrDataNotFound = errors.New("data not found")
	// ErrNoUpdatedData is an error for when no data is provided to update
	ErrNoUpdatedData = errors.New("no data to update")
	// ErrConflictingData is an error for when data conflicts with existing data
	ErrConflictingData = errors.New("data conflicts with existing data in unique column")
	// ErrInvalidCredentials is an error for when the credentials are invalid
	ErrInvalidCredentials = errors.New("invalid email or password")
	// ErrUnauthorized is an error for when the user is unauthorized
	ErrUnauthorized = errors.New("user is unauthorized to access the resource")
)

var errorStatusMap = map[error]int{
        ErrInternal:                   http.StatusInternalServerError,
        ErrDataNotFound:               http.StatusNotFound,
        ErrConflictingData:            http.StatusConflict,
        ErrInvalidCredentials:         http.StatusUnauthorized,
        ErrUnauthorized:               http.StatusUnauthorized,
}

func HandleError(ctx *gin.Context, err error) {
	statusCode, ok := errorStatusMap[err]
	if !ok {
		statusCode = http.StatusInternalServerError
	}

	errRsp := errorResponse(errMsg)
	ctx.JSON(statusCode, errRsp)
}

func errorResponse(errMsgs []string) errorResponse {
	return errorResponse{
		Success:  false,
		Messages: errMsgs,
	}
}
```

File: internal/conf/conf.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	// Cache connection uri
+	// It can be redis format or internal
+	Cache string `envconfig:"CACHE"`
+	// Log level for verbosity
+	LogLevel string `envconfig:"LOG_LEVEL"`
+	// App port
+	Port string `envconfig:"PORT" default:"8000"`
+}
+
+func (c *Config) Validate() error {
```
Comment (anton-sidelnikov) [inline internal/conf/conf.go:27]: move all validations to separate pkg please

**Review by Aloento — CHANGES_REQUESTED** [review]
We should build on V2 to be compatible with the V1 API, not simply move Python code to Golang. If we are just moving code, we can merely use Cython to speed up the original SD2 or further optimize on top of SD2. - **we could do better**.
And this database design, like I said before, can lead to performance issues.

File: internal/app/middleware.go (hunk 2 above)
Comment (Aloento) [reply internal/app/middleware.go:7]: Why not use https://developer.mozilla.org/docs/Web/HTTP/Status ?

**Comment by sgmv** [comment]
> We should build on V2 to be compatible with the V1 API, not simply move Python code to Golang. If we are just moving code, we can merely use Cython to speed up the original SD2 or further optimize on top of SD2. - **we could do better**. And this database design, like I said before, can lead to performance issues.

Don't worry, unfortunately we can't change DB before changing backend, and we should keep backward compatibility first. We will do a lot with schemas a bit later. Please check all endpoints, requests, responses.

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,13 @@
+default: test
```
[the remark this answers is not in the document]
Comment (sgmv) [reply Makefile:1]: Added the SHELL variable, but PATH under discussion :) We will use actions, so maybe we don't need it.

File: .gitignore [hunk 5]
```diff
@@ -0,0 +1,5 @@
+# IDE settings
```
[the remark this answers is not in the document]
Comment (sgmv) [reply .gitignore]: done

File: internal/app/app.go (hunk 1 above)
Comment (sgmv) [reply internal/app/app.go]: Changed.

File: internal/app/middleware.go (hunk 2 above)
Comment (sgmv) [reply internal/app/middleware.go:48]: Yes, the error file will be soon.

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (sgmv) [reply Makefile:1]: So, lint checks work fine without additional PATH. I'll check it with unit tests.

**Comment by Aloento** [comment]
> > We should build on V2 to be compatible with the V1 API, not simply move Python code to Golang. If we are just moving code, we can merely use Cython to speed up the original SD2 or further optimize on top of SD2. - **we could do better**. And this database design, like I said before, can lead to performance issues.
> 
> Don't worry, unfortunately we can't change DB before changing backend, and we should keep backward compatibility first. We will do a lot with schemas a bit later. Please check all endpoints, requests, responses.

just write Database Migration Code ↓

https://github.com/Aloento/StatusDashboard/blob/master/Services/Status.cs#L21

**Comment by Aloento** [comment]
# About the Database

I agree with something Vladi once said, “Later is never.” So, I strongly recommend developing the new database structure now. We don't need to consider compatibility with the old backend because the new backend will run on a new database instance. We only need to perform a one-time data migration when going live, and this process is quite simple. Our test instance can directly use the migrated data for demonstration. The data in my SD3 C# version is migrated in real-time, and it already supports complete and rapid migration. If it’s too complicated to write in Golang, I can modify the C# migration code to adapt to our new backend since migration is a one-time task. I’ve attached my designed database structure, hoping it will help you all.

# OpenAPI v2

Front-end requests are cached, and only a fine-grained API can maximize front-end caching.

## /v2/Event

We do not call it an incident because a maintenance event is not an incident, so now we use Event.

### GET /v2/Events

list all event ids
this is for page https://sd3.eco.tsi-dev.otc-service.com/History

### GET /v2/Event/{id}

get event entity by id
this is for pages like https://sd3.eco.tsi-dev.otc-service.com/Event/205

with related region ids, related component ids
not contain the update data

### GET /v2/Event/{id}/Updates

Return all update entities of the event

### GET /v2/Events/Active

return all Current Events ids
this is for home page component https://sd3.eco.tsi-dev.otc-service.com/#:~:text=working%20on%20it.-,Current%20Events,-ID

### POST /v2/Event

Create a new event

### PATCH /v2/Event/{id}

Update event by patch method and push new update entity at the same time

> The main difference between PUT and PATCH in the REST API is that PUT completes the update operation by replacing the entire entity. In contrast, PATCH only updates the fields you provide and doesn't change any other values. If you use the PUT method, the entire content of the resource you update is updated.

### POST /v2/Event/{id}/Split

by given the [{Service_Id}, {Region_Id}] pairs, split the event into two events.

### POST /v2/Event/Merge

Merge two events into one event by providing the two event IDs, new event titles, descriptions, etc.

## /v2/Region

### GET /v2/Regions

List all region entities

Not contain any component data or event data or category data

### GET /v2/Region/{id}/Categories

Return all categories ids present in the region

## /v2/Category

### GET /v2/Categories

List all categories entity

Not contain any component data or event data

### GET /v2/Category/{id}/Services

Return all component ids present in the category

## /v2/Service

### GET /v2/Services

list all component ids

### GET /v2/Service/{id}

get component entity by id
with related category id, and related region ids

### GET /v2/Service/{id}/Events

Return all event ids present in the component

## /v2/Availability

### GET /v2/Availability/{Service_Id}/{Region_Id}

return the availability of the component in the last 6 months
like [100, 42.24, 70.44, ...]

**Comment by Aloento** [comment]
image

**Review by Aloento — APPROVED** [review]

**Review by anton-sidelnikov — APPROVED** [review]

File: internal/app/routes.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
+package app
+
+const (
+	v1Group = "v1"
```
Comment (Aloento) [inline internal/app/routes.go:4]: `/api/v1` is mandatory

File: openapi.yaml [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+  version: 1.0.0
+servers:
+  - url: http://localhost:8000
+tags:
+  - name: incidents
+    description: Incident management
+  - name: components
+    description: Operations about components
+paths:
+  /v1/components:
```
Comment (Aloento) [inline openapi.yaml:13]: I had thought `/api` was implemented elsewhere, but it's not.
So it has to be modified.

**Pull request merged**